### PR TITLE
Align loss functions with PyTorch's APIs.

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
The most immediate consequence is that losses are modules rather than delegates, which means you need to call `.forward()` to actually compute the loss. Also, the factories are in `torch.nn` rather than `torch.nn.functional` and have the same Pascal-casing as the corresponding types.

The members of the `torch.nn.functional` static class are now proper immediate loss functions, whereas the previous ones returned a loss delegate.

### CODE REVIEW

**Comment by NiklasGustafsson** [comment]
@kaiidams -- here's the loss function fix.

**Review by tarekgh — APPROVED** [review]

**Review by kaiidams — DISMISSED** [review]

File: test/TorchSharpTest/NN.cs [hunk 1]
```diff
@@ -1048,6 +1234,23 @@ public void TestTripleMarginWithDistanceLoss()
...
+                (x, y) => {
+                    return (x - y).abs();
+                };
+
+            using (Tensor anchor = torch.rand(new long[] { 15, 5 }, requiresGrad: true).neg())
+            using (Tensor positive = torch.randn(new long[] { 15, 5 }, requiresGrad: true))
+            using (Tensor negative = torch.randn(new long[] { 15, 5 })) {
+
+                var result = triplet_margin_with_distance_loss(anchor, positive, negative, distance);
+                Assert.True(true);
```
Comment (kaiidams) [inline test/TorchSharpTest/NN.cs:1250]: Probably we could check `Assert.Equal(Array.Empty<long>(), result.shape);` or `Assert.False(float.IsNaN(result.item<float>())` to make sure `result` is a float number.

**Review by kaiidams — APPROVED** [review]

File: test/TorchSharpTest/NN.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply test/TorchSharpTest/NN.cs:1250]: Excellent idea.
